test/val/val_builtins_test.cpp

```diff
@@ -0,0 +1,324 @@
+    %bool    = OpTypeBool
+    %uint32  = OpTypeInt 32 0
+    %float   = OpTypeFloat 32
+    %vecu2   = OpTypeVector %uint32 2
+    %vecu3   = OpTypeVector %uint32 3
+    %vecu4   = OpTypeVector %uint32 4
+    %vecf2   = OpTypeVector %float 2
+    %vecf3   = OpTypeVector %float 3
+    %vecf4   = OpTypeVector %float 4
+    %const2  = OpConstant %uint32 2
+    %const3  = OpConstant %uint32 3
+    %const4  = OpConstant %uint32 4
+    %arru2   = OpTypeArray %uint32 %const2
+    %arru3   = OpTypeArray %uint32 %const3
+    %arru4   = OpTypeArray %uint32 %const4
+    %arrf2   = OpTypeArray %float %const2
+    %arrf3   = OpTypeArray %float %const3
+    %arrf4   = OpTypeArray %float %const4
+    %voidfun = OpTypeFunction %void
+  )";
+  string definitions = R"(
+    %ptr   = OpTypePointer )" + storage + " " + type + R"(
+    %bltin = OpVariable %ptr )" + storage + R"(
+    %entry = OpFunction %void None %voidfun
+    %label = OpLabel
+             OpReturn
+             OpFunctionEnd
+  )";
+
+  return capabilities + extensions + memory + entries
+         + modes + decorations + types + definitions;
+}
+
+// clang-format off
+#define BUILTIN(builtin,type,storage,exec) \
+  make_tuple(string(builtin),string(type),string(storage),string(exec))
+
+TEST_P(ValidateBuiltIn, BuiltInValid) {
+  CompileSuccessfully(GenerateShaderCode(GetParam()));
+  EXPECT_EQ(SPV_SUCCESS, ValidateAndRetrieveValidationState());
+}
+
+INSTANTIATE_TEST_CASE_P(BuiltInValid, ValidateBuiltIn,
+  ::testing::Values(
+       BUILTIN("Position","%vecf4","Output","Vertex"),
+       BUILTIN("PointSize","%float","Output","Vertex"),
+       BUILTIN("ClipDistance","%arrf2","Output","Vertex"),
+       BUILTIN("CullDistance","%arrf3","Output","Vertex"),
+       BUILTIN("VertexId","%uint32","Input","Vertex"),
+       BUILTIN("InstanceId","%float","Input","Vertex"),
+       BUILTIN("PrimitiveId","%uint32","Input","Fragment"),
+       BUILTIN("InvocationId","%uint32","Input","Geometry"),
+       BUILTIN("Layer","%uint32","Input","Fragment"),
+       BUILTIN("ViewportIndex","%uint32","Output","Geometry"),
+       BUILTIN("TessLevelOuter","%arrf4","Output","TessellationControl"),
+       BUILTIN("TessLevelInner","%arrf2","Input","TessellationEvaluation"),
+       BUILTIN("TessCoord","%vecf3","Input","TessellationEvaluation"),
+       BUILTIN("PatchVertices","%uint32","Input","TessellationControl"),
+       BUILTIN("FragCoord","%vecf4","Input","Fragment"),
+       BUILTIN("PointCoord","%vecf2","Input","Fragment"),
+       BUILTIN("FrontFacing","%bool","Input","Fragment"),
+       BUILTIN("SampleId","%uint32","Input","Fragment"),
+       BUILTIN("SamplePosition","%vecf2","Input","Fragment"),
+       BUILTIN("SampleMask","%arru2","Input","Fragment"),
+       BUILTIN("FragDepth","%float","Output","Fragment"),
+       BUILTIN("HelperInvocation","%bool","Input","Fragment"),
+       BUILTIN("NumWorkgroups","%vecu3","Input","GLCompute"),
+       //BUILTIN("WorkgroupSize","%vecu3","Uniform","GLCompute"),
```

Comment:
I didn't know how to treat WorkgroupSize because of the following rule: _"The object decorated with WorkgroupSize must be a specialization constant or a constant."_ I think it requires individual tests